### PR TITLE
Restore model_rebuild calls to top level fragments

### PR DESCRIPTION
**Pull request by bombsimon**
* Restore model_rebuild calls to top level fragments
* Update tests
* Update changelog
* Update example
* Fix `pylint` violation
* Add test to ensure models are rebuilt properly

Superseds #258

### CODE REVIEW

File: tests/main/test_model_rebuild_validation.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+from pydantic_core import ValidationError
+
+from .clients.multiple_fragments.expected_client.example_query_2 import (
+    ExampleQuery2,
+    ExampleQuery2ExampleQuery,
+)
+from .clients.multiple_fragments.expected_client.fragments import FullA
+
+
+def test_model_rebuild_validate():
```
Comment (rafalp) [inline tests/main/test_model_rebuild_validation.py:25]: This test reads to me like a four tests bunched together. It should be split into those.

**Review by rafalp — CHANGES_REQUESTED** [review]
LGTM but we could improve the test.

File: tests/main/test_model_rebuild_validation.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+"""
+To ensure all models with nested dependencies are fully rebuilt this test will
```
[the remark this answers is not in the document]
Comment (bombsimon) [reply tests/main/test_model_rebuild_validation.py]: Updated!

File: tests/main/test_model_rebuild_validation.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    ExampleQuery2ExampleQuery,
+)
+from .clients.multiple_fragments.expected_client.fragments import FullA
+
+
+def test_model_rebuild_validate():
+    # Perform some sanity checks on the schema for `ExampleQuery2` to test that
+    # it confirms to the fields of `FullA` and that it references `FullB`.
+    json_schema = ExampleQuery2.model_json_schema()
+    assert all(
```
[the remark this answers is not in the document]
Comment (bombsimon) [reply tests/main/test_model_rebuild_validation.py]: Makes sense, split them into separate assertions!

File: tests/main/test_model_rebuild_validation.py (hunk 1 above)
Comment (bombsimon) [reply tests/main/test_model_rebuild_validation.py]: Made a fixture and split the test into 4 different steps which also made us able to remove the comments, good idea! 👍

File: tests/main/test_model_rebuild_validation.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+"""
+To ensure all models with nested dependencies are fully rebuilt this test
```
Comment (rafalp) [inline tests/main/test_model_rebuild_validation.py:2]:
```suggestion
To ensure all models with nested dependencies are fully rebuilt those tests
```

File: tests/main/test_model_rebuild_validation.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+creates an instance of the query from `multiple_fragments` containing the
```
Comment (rafalp) [inline tests/main/test_model_rebuild_validation.py:3]:
```suggestion
create an instance of the query from `multiple_fragments` containing the
```

File: tests/main/test_model_rebuild_validation.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+`FullA` fragment (used by the `ExampleQuery2ExampleQuery`) which itself includes
+a field of type `FullAFieldB` that extends the `FullB` fragment.
+
+If this model is not rebuilt with `FullA.model_rebuild()` `ExampleQuery2` will
+not be fully defined and we will get a `PydanticUserError`.
```
Comment (rafalp) [inline tests/main/test_model_rebuild_validation.py:8]:
```suggestion
not be fully defined and we will raise a `PydanticUserError`.
```

**Review by rafalp — CHANGES_REQUESTED** [review]

File: tests/main/test_model_rebuild_validation.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    query_props = json_schema["$defs"]["ExampleQuery2ExampleQuery"]["properties"]
+    assert "id" in query_props
+    assert "value" in query_props
+    assert "fieldB" in query_props
+    assert query_props["fieldB"]["$ref"] == "#/$defs/FullAFieldB"
+
+
+@pytest.fixture
+def field_a():
```
Comment (rafalp) [inline tests/main/test_model_rebuild_validation.py:38]:
```suggestion
def field_a_data():
```

**Review by rafalp — CHANGES_REQUESTED** [review]

File: tests/main/test_model_rebuild_validation.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+def test_validate_field_a_on_faulty_model(field_a_data):
+    with pytest.raises(ValidationError):
+        ExampleQuery2.model_validate(field_a_data)
+
+
+def test_validate_field_a_on_correct_model(field_a_data):
+    try:
+        FullA.model_validate(field_a_data)
+        ExampleQuery2ExampleQuery.model_validate(field_a_data)
+    except ValidationError as e:
```
Comment (rafalp) [inline tests/main/test_model_rebuild_validation.py:54]: Lets drop those extra try/except with failing asserts, they are adding noise to tests.

**Review by rafalp — CHANGES_REQUESTED** [review]

**Review by rafalp — APPROVED** [review]

**Pull request merged**